### PR TITLE
make bn shareable when frozen

### PR DESCRIPTION
**Pull request by HoneyryderChuck**
added frozen check on every state change.

### CODE REVIEW

**Comment by rhenium** [comment]
Could you make the part that changes `#include <ruby/ractor.h>` a separate commit or PR?

**Comment by HoneyryderChuck** [comment]
@rhenium [done](https://github.com/ruby/openssl/pull/811). I'll rebase it here once it's merged.

File: ext/openssl/ossl_bn.c [hunk 1]
```diff
@@ -748,6 +745,7 @@ ossl_bn_is_bit_set(VALUE self, VALUE bit)
 	BIGNUM *bn, *result;				\
 	int b;						\
 	VALUE obj;					\
+	rb_check_frozen(self);			\
```
[the remark this answers is not in the document]
Comment (HoneyryderChuck) [reply ext/openssl/ossl_bn.c]: true, removed.

**Comment by HoneyryderChuck** [comment]
@rhenium  done

**Review by rhenium — CHANGES_REQUESTED** [review]
`#initialize` also needs a frozen check.

Could you add a test case to ensure freezing it actually makes it shareable?

File: ext/openssl/ossl_bn.c [hunk 2]
```diff
@@ -689,6 +689,7 @@ BIGNUM_3c(mod_exp)
     ossl_bn_##func(VALUE self, VALUE bit)		\
     {							\
 	BIGNUM *bn;					\
+	rb_check_frozen(self);			\
```
[the remark this answers is not in the document]
Comment (HoneyryderChuck) [reply ext/openssl/ossl_bn.c]: done

**Comment by HoneyryderChuck** [comment]
> #initialize also needs a frozen check.

why and where? 🤔 

added a test.

**Comment by rhenium** [comment]
`ossl_bn_initialize()` - it doesn't have a `rb_check_frozen(self)`.

**Comment by HoneyryderChuck** [comment]
gotcha, added. One question though: is it worth it, considering that on intialize, the object isn't expected to be frozen? What would that guard against?

**Comment by rhenium** [comment]
It's needed because it's callable from Ruby, especially if the object is now expected to be thread safe when frozen. Whether calling `#initialize` on a frozen object makes sense or not doesn't matter.

**Review by rhenium — APPROVED** [review]

**Comment by rhenium** [comment]
It looks good to me now. Thanks!

**Pull request merged**